frontend-react/docs/proposals/0002-domain-driven-directory-structure.md

```diff
@@ -0,0 +1,148 @@
+# 0002: Domain-Driven Directory Structure
+
+* Status: proposed
+* Authors: @stephenkao
+* Deciders: [list everyone involved in the decision]
+* Proposal date: 2023-01-12
+* Last updated: 2023-01-12
+
+## Problem Statement
+
+Currently, the `frontend-react/` directory has a flat structure, meaning that we have one `components/` directory for components, one `utils/` directory for utilities, one `hooks/` directory for hooks, et cetera.  This works for a simple application, but as we continue to build new features and add new files, it's becoming more difficult to immediately answer questions about where specific files should live and how we should separate logic into said files.
+
+## Decision Drivers
+
+- Lack of teamwide alignment on when we should create new files and where files should live
+- Consolidation of a component system for deviations from/customizations on top of USWDS
+- Desire for predictability, colocation, and ease of searching within directories
+
+## Considered Options
+
+### Option 1: Continue with current directory structure
+
+#### Description
+
+This option proposes no change to our current organization so the structure will remain as-is.
+
+```
+src/
+  components/
+    AuthElement.tsx
+    AuthElement.test.tsx
+    Admin/
+      AdminFormEdit.test.tsx
+      AdminFormEdit.tsx
+    alerts/
+      NoServicesAlert.test.tsx
+      NoServicesAlert.tsx
+  utils/
+    Analytics.test.ts
+    Analytics.ts
+  contexts/
+  hooks/
+  styles/
+```
+
+#### Pros
+
+- No time or effort needed
+- No risk of file migration
+
+#### Cons
+- Harder to maintain
+- No consolidation on where files should live
+- Opaque colocation of related logic
+- Lack of consistency in naming, subdirectories
+- One broad top-level directory for each of components, hooks, styles, utilities, all of which are only similar in type, not purpose
+
+### Option 2: Introduce domain-driven directory structure
+
+#### Description
+
+This option proposes we group our files by domain, a classification by functionally related logic, and each domain will have its own directory with the following domain units:
+
+- `components`: all domain-specific React component files, including component definitions, tests, colocated styles, and Storybook stories
+- `hooks`: all domain-specific React hooks and tests
+- `decorators`: all domain-specific React decorators
+- `utils`: all domain-specific utilities, tests, and endpoint resources
+
+A typical structure will look like the following:
+
+```
+src/
+  <domain>/
+    components/
+      ComponentA.tsx
```

Comment:
One gripe I have with this is that it feels like it'll step on code reuse. When I consider React as a framework, I think of `components` and `hooks` being two unique "tools" to build with, not specific to any single domain on the site, like a page or feature. Consider building an IKEA shelf vs hanging a picture frame. Both require a hammer, one requires nails, and one requires dowels (that IKEA always so kindly gives you one less of than is required, but I digress). Shelves and pictures can live anywhere in a house, though...do we own a "living room hammer" and a "bedroom hammer"? No, because that's silly!

The nature of React's tools is reusability, and it gets hard to isolate where something is and whether it's utility applies elsewhere when it's domain-specific. Taking that hammer analogy one step further, there's many types of hammers to do many jobs, but the core of each hammer is the same: handle, head. If we abstracted things from each domain into an "abstract" domain, and build from each of those, we run the risk of over-abstraction (at least in my experience). 

I'd propose something more akin to this:

```
src/
    components/
        ... (sub directories as they make sense)
    hooks/
        ... (sub directories as they make sense)
    app/
        <domain>/
            DomainPage.tsx
            ...
``` 

This would still maintain these core tools of React as reusable pieces of code rather than one domain borrowing from another domain, and give us the ability to further consider what _types_ of "hammers" we need, so to speak, while keeping domain-specific code as the least reusable part of the front-end codebase, like a page that composes hooks and components in such a way that is _only_ meant for use at a specific route.

Reply:
And as a quick follow up, for an example: consider what'd happen if we locked hooks into a domain, but that domain - say, an unused page/feature - no longer gets used. When it comes time to delete that domain from the repository, that subdirectory is tied into other parts of the codebase if we've reused HookA or ComponentB somewhere in the application other than that directory.

That's mainly what I mean when I said (then edited out) "it steps on code reusability". I changed the wording to fit my point above better to "it sends the message that reusability isn't our goal", but in this example, I think "steps on" is much more apt.

Reply:
I don't think anyone's arguing against reusability in our codebase, just that the way that our files are organized right now isn't going to do us any favors as we continue building in it.  One of the benefits for a domain-driven folder structure is for the ease of colocation and maintenance in understanding that a file in any particular domain is _likely_ to used in or affect another file in the same domain.

Domains are intended to operate as separate modules, yes, but outside of the `common`/`core`/`shared` domain, there realistically should be minimal reuse between domains; anything generic or atomic enough should be elevated to said domain.  Working in a larger React app means that we're going to be building a lot of "unique" components/hooks frequently, but if they're all built on top of the same atomic parts, they're just going to be consumers of the same underlying logic rather than producing anything novel.  (Just look at  `Admin/`, `header/`, and `MessageTracker/` as an example of this.)  And if we find ourselves repeating the same patterns over and over again, then in my opinion that'll warrant a cleanup ticket to escalate into a `common`/`core`/`shared` file.

My feeling here is that real reusability isn't something that can be enforced through a directory structure; there's nothing stopping engineers from reinventing the wheel ad nauseam.  (This is how we ended up with the NPM community.)  Reusability is something we need to keep top of mind as we're building out our core component system and making the code's capabilities known to design/product during planning.  Instead, the directory structure can help to build a mental model of where certain logic lives and to enforce better patterns around encapsulation.

Reply:
>Domains are intended to operate as separate modules, yes, but outside of the common/core/shared domain, there realistically should be minimal reuse between domains

Ahh, I think I misunderstood domains as "partitioned areas of the site" not "partitioned areas of functionality" from the way the file structure was shaped. Common, core, and shared domains are exactly what I'd considered the "as it makes sense" subdirectories of hooks and components. 

>My feeling here is that real reusability isn't something that can be enforced through a directory structure; there's nothing stopping engineers from reinventing the wheel ad nauseam.

Agreed, though my statement was leaning more towards the philosophical approach; sort of a question of "what are we saying _without_ words when we display code segmented by domain-specific partitions (again, under the assumption that a "common/core/shared" domain didn't exist in this file structure). I think those non-verbal comms are what help newcomers to a project see our intent.

Reply:
Yeah, exactly!  Domains are more about consolidation for evergreen concepts like authentication, Organizations, Senders/Receivers, admin stuff...things that won't be going away anytime soon unless ReportStream dramatically changes its mission statement -- at which point...we may just want to start an entirely new codebase 😆

I'll update the proposal to be a little clearer about this!